arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,755 @@
+	 *
+	 * Remember that the boundary of each memory region is marked by
+	 * two consecutive entries, and that the addresses of all entries
+	 * must not be in descending order (i.e. equal or increasing).
+	 * To ensure this, we clear out the entries first then write them
+	 * in reverse order. This avoids any intermediate invalid
+	 * configuration with regard to ordering.
+	 */
+	for (entry = 0; entry < XTENSA_MPU_NUM_ENTRIES; entry++) {
+		__asm__ volatile("wptlb %0, %1\n\t" : : "a"(entry), "a"(0));
+	}
+
+	for (entry = XTENSA_MPU_NUM_ENTRIES - 1; entry >= 0; entry--) {
+		__asm__ volatile("wptlb %0, %1\n\t"
+				 : : "a"(map->entries[entry].at), "a"(map->entries[entry].as));
+	}
+}
+
+/**
+ * Perform necessary steps to enable MPU.
+ */
+void xtensa_mpu_init(void)
+{
+	unsigned int entry;
+	uint8_t first_enabled_idx = XTENSA_MPU_NUM_ENTRIES;
+
+	/* Disable all foreground segments before we start configuration. */
+	xtensa_mpu_mpuenb_write(0);
+
+	/*
+	 * Clear the foreground MPU map so we can populate it later with valid entries.
+	 * Note that we still need to make sure the map is valid, and cannot be totally
+	 * zeroed.
+	 */
+	for (entry = 0; entry < XTENSA_MPU_NUM_ENTRIES; entry++) {
+		/* Make sure MBZ fields are cleared. */
+		xtensa_mpu_map_fg_kernel.entries[entry].as.raw = 0U;
+		xtensa_mpu_map_fg_kernel.entries[entry].at.raw = 0U;
+
+		/* Segment value must correspond to the index. */
+		xtensa_mpu_map_fg_kernel.entries[entry].at.p.segment = entry;
+
+		/* No access at all for both kernel and user modes. */
+		xtensa_mpu_map_fg_kernel.entries[entry].at.p.access_rights = 0;
+
+		/*
+		 * According to manual, theis is the default memory type for
+		 * background map. So use it here to avoid invalid value.
+		 */
+		xtensa_mpu_map_fg_kernel.entries[entry].at.p.memory_type = 0x06U;
+	}
+
+	/*
+	 * Add necessary MPU entries for the memory regions of base Zephyr image.
+	 */
+	for (entry = 0; entry < ARRAY_SIZE(mpu_zephyr_ranges); entry++) {
+		const struct xtensa_mpu_range *range = &mpu_zephyr_ranges[entry];
+		uint16_t memory_type = xtensa_mpu_memory_flags_to_type(range->memory_flags);
+
+		int ret = mpu_map_region_add(&xtensa_mpu_map_fg_kernel,
+					     range->start, range->end,
+					     range->access_rights, memory_type,
+					     &first_enabled_idx);
+
+		ARG_UNUSED(ret);
+		__ASSERT(ret == 0, "Unable to add region [0x%08x, 0x%08x): %d",
+				   (unsigned int)range->start,
+				   (unsigned int)range->end,
+				   ret);
+	}
+
+#if defined(__GNUC__)
+	/**
+	 * GCC complains about usage of the SoC MPU range ARRAY_SIZE
+	 * (xtensa_soc_mpu_ranges) as the default weak declaration is
+	 * an empty array, and any access to its element is considered
+	 * out of bound access. However, we have a number of element
+	 * variable to guard against this (... if done correctly).
+	 * Besides, this will almost be overridden by the SoC layer.
+	 * So tell GCC to ignore this.
+	 */
```

Comment:
I remember this comment from somewhere else too.  While this works, dealing with compiler pragmas seems needless.  Just declare the symbol extern here and put the __weak definition in another C file where it doesn't get accessed.

Reply:
I turned the whole thing into "required" so no more weak symbol and this is no longer needed.